### PR TITLE
[Perf] Mem align KV caches for CUDA devices (MLA perf improvement)

### PR DESCRIPTION
**Pull request by LucasWilkinson**
Generally Nvidia hardware likes 256 byte alignment (reasons is foggy due to the blackbox nature of Nvidia hardware), but memory allocated via the CUDA Runtime ensure 256 byte alignment (see https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/#a-sequential-but-misaligned-access-pattern).

This PR aligns KV cache entries to start 256 byte boundaries, this mainly targets MLA since for "normal attention" with normal head dims (say 64 or 128) the entries are naturally 256 byte aligned.

This means for MLA with a head dim of 576 (like DeepSeek V2/V3) and a fp16/bf16 cache, we allocate 640 elements per cache entry in instead of 576 (1280 bytes instead of 1152). This increases the size of the cache by ~11% (wasted), but leads to a worthwhile performance gain.

Results DeepSeek-R1 on 8xH200

```
VLLM_CUDA_MEM_ALIGN_KV_CACHE=0  python3 benchmarks/benchmark_throughput.py --model /data/nm/models/DeepSeek-R1 --trust-remote-code --tensor-parallel-size 8 --max-model-len 8000 --enable-chunked-prefill False --input-len 2000 --output-len 1000  --num-prompts 100
...
Throughput: 0.76 requests/s, 2289.10 total tokens/s, 763.03 output tokens/s
```

```
VLLM_CUDA_MEM_ALIGN_KV_CACHE=1  python3 benchmarks/benchmark_throughput.py --model /data/nm/models/DeepSeek-R1 --trust-remote-code --tensor-parallel-size 8 --max-model-len 8000 --enable-chunked-prefill False --input-len 2000 --output-len 1000  --num-prompts 100
...
Throughput: 1.10 requests/s, 3287.09 total tokens/s, 1095.70 output tokens/s
```

Accuracy:
```
VLLM_MLA_DISABLE=1 lm_eval --model vllm --model_args pretrained=/data/nm/models/DeepSeek-R1,tensor_parallel_size=8,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=False --task gsm8k --num_fewshot=5 --limit 100
...
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.94|±  |0.0239|
|     |       |strict-match    |     5|exact_match|↑  | 0.94|±  |0.0239|


VLLM_MLA_DISABLE=0 VLLM_CUDA_MEM_ALIGN_KV_CACHE=0 lm_eval --model vllm --model_args pretrained=/data/nm/models/DeepSeek-R1,tensor_parallel_size=8,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=False --task gsm8k --num_fewshot=5 --limit 100
...
INFO 02-03 14:26:12 executor_base.py:110] # CUDA blocks: 30218, # CPU blocks: 3819
INFO 02-03 14:26:12 executor_base.py:115] Maximum concurrency for 16384 tokens per request: 29.51x
...
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.94|±  |0.0239|
|     |       |strict-match    |     5|exact_match|↑  | 0.94|±  |0.0239|


VLLM_MLA_DISABLE=0 VLLM_CUDA_MEM_ALIGN_KV_CACHE=1 lm_eval --model vllm --model_args pretrained=/data/nm/models/DeepSeek-R1,tensor_parallel_size=8,dtype=auto,gpu_memory_utilization=0.9,trust_remote_code=True,max_model_len=16384,enforce_eager=False --task gsm8k --num_fewshot=5 --limit 100
...
INFO 02-03 14:33:20 executor_base.py:110] # CUDA blocks: 27196, # CPU blocks: 3437
INFO 02-03 14:33:20 executor_base.py:115] Maximum concurrency for 16384 tokens per request: 26.56x
...
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.94|±  |0.0239|
|     |       |strict-match    |     5|exact_match|↑  | 0.94|±  |0.0239|
```

### CODE REVIEW

File: vllm/worker/cache_engine.py [hunk 1]
```diff
@@ -110,14 +130,20 @@ def get_cache_block_size(
         num_attention_layers = model_config.get_num_layers_by_block_type(
             parallel_config, LayerBlockType.attention)
 
-        key_cache_block = cache_config.block_size * num_heads * head_size
```
Comment (tlrmchlsmth) [inline vllm/worker/cache_engine.py:113]: does this line need to be restored, given the current pre-commit error "Undefined name `key_cache_block`"?

File: vllm/worker/cache_engine.py [hunk 2]
```diff
@@ -75,15 +80,30 @@ def _allocate_kv_cache(
             num_blocks, self.block_size, self.num_kv_heads, self.head_size)
         pin_memory = is_pin_memory_available() if device == "cpu" else False
         kv_cache: List[torch.Tensor] = []
+        
+        entry_shape = kv_cache_shape[2:]
+        entry_size = np.prod(entry_shape)
+        
+        # Align entrys so they are 256 byte aligned for better performance
```
Comment (tlrmchlsmth) [inline vllm/worker/cache_engine.py:87]: nit: entries

File: vllm/envs.py [hunk 3]
```diff
@@ -539,6 +540,15 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
     "VLLM_ENABLE_MOE_ALIGN_BLOCK_SIZE_TRITON":
     lambda: bool(int(os.getenv("VLLM_ENABLE_MOE_ALIGN_BLOCK_SIZE_TRITON", "0"))
                  ),
+
+    # When on a Nvidia GPU aligns single entrys (within a page) so they are 256
```
Comment (tlrmchlsmth) [inline vllm/envs.py:544]: nit: entrys -> entries

File: vllm/envs.py [hunk 4]
```diff
@@ -539,6 +540,15 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
...
+    # byte aligned for  better performance, this increases the memory usage of 
+    # the cache. Currently this primarily affects MLA that results in non-256
+    # byte aligned entrys. This mathches the alginment the CUDA runtime uses
```
Comment (tlrmchlsmth) [inline vllm/envs.py:547]: nit: entrys -> entires, mathches -> matches, alginment -> alignment

File: vllm/worker/cache_engine.py [hunk 5]
```diff
@@ -75,15 +80,30 @@ def _allocate_kv_cache(
...
-                            device=device))
+            layer_kv_cache = torch.zeros(alloc_shape,
+                                         dtype=self.dtype,
+                                         pin_memory=pin_memory,
+                                         device=device)
+            
+            if alloc_entry_size != entry_size:
+                layer_kv_cache = layer_kv_cache[..., :entry_size]
+
+            kv_cache.append(layer_kv_cache.view(kv_cache_shape))
```
Comment (tlrmchlsmth) [inline vllm/worker/cache_engine.py:106]: The implementation looks good to me. A couple of comments noting what the padding and views are doing would be nice to make it a little easier to follow (as well as noting that this is kind of a special case for MLA).

**Review by tlrmchlsmth — APPROVED** [review]
Great find! Makes a lot of sense!

File: vllm/envs.py [hunk 6]
```diff
@@ -539,6 +540,15 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
     "VLLM_ENABLE_MOE_ALIGN_BLOCK_SIZE_TRITON":
     lambda: bool(int(os.getenv("VLLM_ENABLE_MOE_ALIGN_BLOCK_SIZE_TRITON", "0"))
                  ),
+
+    # When on a Nvidia GPU aligns single entrys (within a page) so they are 256
+    # byte aligned for  better performance, this increases the memory usage of
```
Comment (mgoin) [inline vllm/envs.py:545]:
```suggestion
    # byte aligned for better performance, this increases the memory usage of 
```

File: vllm/envs.py [hunk 7]
```diff
@@ -539,6 +540,15 @@ def maybe_convert_int(value: Optional[str]) -> Optional[int]:
...
                  ),
+
+    # When on a Nvidia GPU aligns single entrys (within a page) so they are 256
+    # byte aligned for  better performance, this increases the memory usage of 
+    # the cache. Currently this primarily affects MLA that results in non-256
+    # byte aligned entrys. This mathches the alginment the CUDA runtime uses
+    # for all allocations. Currently this primarily affects MLA, for most other
+    # models the alignment is already naturally aligned to 256 bytes.
+    "VLLM_CUDA_MEM_ALIGN_KV_CACHE":
+    lambda: bool(int(os.getenv("VLLM_CUDA_MEM_ALIGN_KV_CACHE", "1"))),
```
Comment (simon-mo) [inline vllm/envs.py:551]: Do we need to flag this? I think we can just turn this on by default?

Comment (mgoin) [reply vllm/envs.py:551]: I agree with the concern about reducing the cache space by ~11%, although maybe we consider this change necessary for performance to remove the choice like you say. 

> This means for MLA with a head dim of 576 (like DeepSeek V2/V3) and a fp16/bf16 cache, we allocate 640 elements per cache entry in instead of 576 (1280 bytes instead of 1152). This increases the size of the cache by ~11% (wasted), but leads to a worthwhile performance gain.

Comment (LucasWilkinson) [reply vllm/envs.py:551]: Given that it can increase the size of the KV-cache I wanted it on by default but with a flag to turn it off incase a user really wants to maximize KV-cache size

File: vllm/_custom_ops.py [hunk 8]
```diff
@@ -1037,6 +1037,11 @@ def copy_blocks(key_caches: List[torch.Tensor],
     torch.ops._C_cache_ops.copy_blocks(key_caches, value_caches, block_mapping)
 
 
+def copy_blocks_mla(kv_caches: List[torch.Tensor],
+                    block_mapping: torch.Tensor) -> None:
+    torch.ops._C_cache_ops.copy_blocks_mla(kv_caches, block_mapping)
```
Comment (tlrmchlsmth) [inline vllm/_custom_ops.py:1042]: Why didn't we need this kernel before?

Comment (LucasWilkinson) [reply vllm/_custom_ops.py:1042]: I think we did ..., I think this may solve some bugs (TBH im not sure how copy_blocks is used by the wider system)

**Review by tlrmchlsmth — APPROVED** [review]
Great work -- LGTM!

**Review by mgoin — APPROVED** [review]

**Comment by leepoly** [comment]
Hi, I’m using a 8xH200 setup and unable to reproduce the results from benchmarks/benchmark_throughput.py in the PR.

My results:
With VLLM_CUDA_MEM_ALIGN_KV_CACHE=0:
- CUDA blocks: 34,497
- Throughput: 1.06 requests/s, 3,170.77 total tokens/s, 1,056.92 output tokens/s

With VLLM_CUDA_MEM_ALIGN_KV_CACHE=1:
- CUDA blocks: 31,047
- Throughput: 1.06 requests/s, 3,185.99 total tokens/s, 1,062.00 output tokens/s

The throughput seems nearly identical in both cases. Could you suggest potential causes for this discrepancy?

**Comment by LucasWilkinson** [comment]
@leepoly what model is this? this only affects MLA (i.e. DeepSeek V2/3)

**Comment by leepoly** [comment]
> @leepoly what model is this? this only affects MLA (i.e. DeepSeek V2/3)
> 
> Edit: nvm I assume you are using R1 since the numbers look very comparable, ill try to re-run the numbers tmrw to see if there is something weird going on

Yes I use deepseek v3 model. And I simply used the script you provided `VLLM_CUDA_MEM_ALIGN_KV_CACHE=0  python3 benchmarks/benchmark_throughput.py --model /data/nm/models/DeepSeek-R1 --trust-remote-code --tensor-parallel-size 8 --max-model-len 8000 --enable-chunked-prefill False --input-len 2000 --output-len 1000  --num-prompts 100`

Even with VLLM_CUDA_MEM_ALIGN_KV_CACHE=0 the reported throughput (1.06rps) already roughly matches with your results with 256B aligned (1.10 rps).
